pippy/PipelineSchedule.py

```diff
@@ -719,31 +719,30 @@ def backward_stage_local_index(step):
                     bwd_mb_index == self._n_microbatches - 1
                 )
                 logger.debug(
-                    f"{self.rank}: {step=}, {fwd_stage.stage_index=}, {bwd_stage.stage_index=}, {fwd_mb_index=}, {bwd_mb_index=}"
+                    f"Rank {self.rank}: {step=}, {fwd_stage.stage_index=}, {bwd_stage.stage_index=}, {fwd_mb_index=}, {bwd_mb_index=}"
                 )
                 with record_function(f"1F1B {step}"):
-                    ops = fwd_stage.get_fwd_recv_ops()
+                    ops.extend(fwd_stage.get_fwd_recv_ops())
                     ops.extend(bwd_stage.get_bwd_recv_ops())
-                    works = sorted_batch_isend_irecv(ops)
-                    for work in works.values():
+                    if ops:
+                        work = dist.batch_isend_irecv(ops).pop()
                         work.wait()
+                        sends_to_wait.append(work)
+                        ops.clear()
 
-                    # fwd
+                    # Forward
                     output = fwd_stage.forward_one_chunk(arg_mbs[fwd_mb_index], kwarg_mbs[fwd_mb_index])  # type: ignore[index]
-                    ops = fwd_stage.get_fwd_send_ops()
+                    ops.extend(fwd_stage.get_fwd_send_ops())
                     self._maybe_compute_loss(
                         fwd_stage, output, target_mbs, fwd_mb_index
                     )
 
-                    # bwd
+                    # Backward
                     loss = self._maybe_get_loss(bwd_stage, bwd_mb_index)
                     bwd_stage.backward_one_chunk(loss=loss)
                     ops.extend(bwd_stage.get_bwd_send_ops())
 
-                    works = sorted_batch_isend_irecv(ops)
-                    sends_to_wait.extend(works.values())
```

Comment:
I assume this firing get merged with the firing right above?

Reply:
yes it gets executed on the next step